### PR TITLE
Add the ability to pass session to test helpers

### PR DESCRIPTION
**Pull request by masad-frost**
# Why

Some services need to do something different based on who the message is from.

# What changed

Added the ability to pass in a session to the different procedures

# Test plan

The follwing test passes, didn't commit it but maybe we should treat `testHelpers` as part of the public API and add tests for the helpers. Lmk what you think.

```ts

  test('asProcedure with session', async () => {
    const BroadCastService = ServiceSchema.define(
      {
        initializeState: () => ({
          clients: new Map<string, (msg: string) => void>(),
        }),
      },
      {
        start: Procedure.stream({
          input: Type.String(),
          output: Type.String(),
          async handler(ctx, input, output) {
            ctx.state.clients.set(ctx.from, (msg) => output.push(Ok(msg)));

            for await (const msg of input) {
              for (const [clientId, send] of ctx.state.clients.entries()) {
                if (clientId === ctx.from) {
                  continue;
                }

                send(msg);
              }
            }
          },
        }),
      },
    );

    const service = BroadCastService.instantiate();
    const state = {
      clients: new Map(),
    };
    const [clientFooInput, clientFooOutput] = asClientStream(
      state,
      service.procedures.start,
      null,
      undefined,
      new Session<Connection>(
        undefined,
        'foo',
        'server',
        testingSessionOptions,
      ),
    );

    const [clientBarInput, clientBarOuput] = asClientStream(
      state,
      service.procedures.start,
      null,
      undefined,
      new Session<Connection>(
        undefined,
        'bar',
        'server',
        testingSessionOptions,
      ),
    );

    clientBarInput.push('hello');
    expect((await iterNext(clientFooOutput)).payload).toStrictEqual('hello');
    expect(clientBarOuput.readableLength).toBe(0);

    clientFooInput.push('world');
    expect((await iterNext(clientBarOuput)).payload).toStrictEqual('world');
    expect(clientFooOutput.readableLength).toBe(0);
  });
```

Here's a patch:

```patch
diff --git a/__tests__/handler.test.ts b/__tests__/handler.test.ts
index 3f650c9..b8790ae 100644
--- a/__tests__/handler.test.ts
+++ b/__tests__/handler.test.ts
@@ -4,6 +4,7 @@ import {
   asClientSubscription,
   asClientUpload,
   iterNext,
+  testingSessionOptions,
 } from '../util/testHelpers';
 import { assert, describe, expect, test } from 'vitest';
 import {
@@ -14,8 +15,11 @@ import {
   SubscribableServiceSchema,
   UploadableServiceSchema,
 } from './fixtures/services';
-import { UNCAUGHT_ERROR } from '../router/result';
+import { Ok, UNCAUGHT_ERROR } from '../router/result';
 import { Observable } from './fixtures/observable';
+import { Procedure, ServiceSchema } from '../router';
+import { Type } from '@sinclair/typebox';
+import { Connection, Session } from '../transport';
 
 describe('server-side test', () => {
   const service = TestServiceSchema.instantiate();
@@ -169,4 +173,71 @@ describe('server-side test', () => {
       payload: { result: 'test 3' },
     });
   });
+
+  test('asProcedure with session', async () => {
+    const BroadCastService = ServiceSchema.define(
+      {
+        initializeState: () => ({
+          clients: new Map<string, (msg: string) => void>(),
+        }),
+      },
+      {
+        start: Procedure.stream({
+          input: Type.String(),
+          output: Type.String(),
+          async handler(ctx, input, output) {
+            ctx.state.clients.set(ctx.from, (msg) => output.push(Ok(msg)));
+
+            for await (const msg of input) {
+              for (const [clientId, send] of ctx.state.clients.entries()) {
+                if (clientId === ctx.from) {
+                  continue;
+                }
+
+                send(msg);
+              }
+            }
+          },
+        }),
+      },
+    );
+
+    const service = BroadCastService.instantiate();
+    const state = {
+      clients: new Map(),
+    };
+    const [clientFooInput, clientFooOutput] = asClientStream(
+      state,
+      service.procedures.start,
+      null,
+      undefined,
+      new Session<Connection>(
+        undefined,
+        'foo',
+        'server',
+        testingSessionOptions,
+      ),
+    );
+
+    const [clientBarInput, clientBarOuput] = asClientStream(
+      state,
+      service.procedures.start,
+      null,
+      undefined,
+      new Session<Connection>(
+        undefined,
+        'bar',
+        'server',
+        testingSessionOptions,
+      ),
+    );
+
+    clientBarInput.push('hello');
+    expect((await iterNext(clientFooOutput)).payload).toStrictEqual('hello');
+    expect(clientBarOuput.readableLength).toBe(0);
+
+    clientFooInput.push('world');
+    expect((await iterNext(clientBarOuput)).payload).toStrictEqual('world');
+    expect(clientFooOutput.readableLength).toBe(0);
+  });
 });
```

### CODE REVIEW

File: util/testHelpers.ts [hunk 1]
```diff
@@ -146,22 +146,25 @@ function catchProcError(err: unknown) {
 
 export const testingSessionOptions: SessionOptions = defaultTransportOptions;
 
-function dummyCtx<State>(
-  state: State,
-  extendedContext?: Omit<ServiceContext, 'state'>,
-): ServiceContextWithTransportInfo<State> {
-  const session = new Session<Connection>(
+function dummySession() {
```
Comment (jackyzha0) [inline util/testHelpers.ts:149]: should we provide the ability to override `to` and `from`?

File: util/testHelpers.ts [hunk 2]
```diff
@@ -146,22 +146,25 @@ function catchProcError(err: unknown) {
...
+  return new Session<Connection>(
     undefined,
     'client',
-    'SERVER',
+    'server',
```
Comment (jackyzha0) [inline util/testHelpers.ts:153]: why change the capitalization here?

**Review by jackyzha0 — APPROVED** [review]
test the test helpers :pog:

not opposed to it at all, go for it

File: util/testHelpers.ts (hunk 1 above)
Comment (masad-frost) [reply util/testHelpers.ts:149]: Should we expose `dummySession` as a helper, I thought people can just import `Session` and do it themselves. I added this method just to dry it up below in the defaults.

**Review by siliconjungle — APPROVED** [review]

File: util/testHelpers.ts (hunk 2 above)
Comment (masad-frost) [reply util/testHelpers.ts:153]: oops

File: util/testHelpers.ts (hunk 1 above)
Comment (jackyzha0) [reply util/testHelpers.ts:149]: session itself isn't that ergonomic to construct so thats why i thought about it but no strong opinions

**Comment by masad-frost** [comment]
Gonna hold off committing the test, we have enough coverage internally, can revisit soon.

**Pull request merged**